fdt.cc

```diff
@@ -1414,13 +1423,38 @@ device_tree::resolve_cross_references(uint32_t &phandle)
 				return;
 			}
 		}
+		// Track referenced nodes for later garbage collection
+		referenced_nodes.push_back(target);
 		// If there is an existing phandle, use it
 		property_ptr p = assign_phandle(target, phandle);
 		p->begin()->push_to_buffer(i.get().val.byte_data);
 		assert(i.get().val.byte_data.size() == 4);
 	}
 }
 
+void
+device_tree::garbage_collect_marked_nodes(node_ptr &n, node *parent)
+{
+	if (n->delete_if_unreferenced)
+	{
+		auto i = std::find(referenced_nodes.begin(), referenced_nodes.end(),
+			n.get());
+		if (i == referenced_nodes.end())
+		{
+			parent->delete_child(n);
+			// No point in checking its children; it has been deleted.
+			return;
+		}
+	}
+	for (auto &c : n->child_nodes())
+	{
+		if (c)
```

Comment:
I think this might be wrong, but it worked. I was getting segfaults, presumably because the iterator gets invalidated when we recurse into children and find one that needs deleted. It seems to somehow recover and properly delete subsequent children that need deleted, but should I instead rewrite this into two different functions, instead collecting a list of the children that need to be removed and .erase() them all at once?

Reply:
Yup, deletion will invalidate the iterators.  You should push the to-delete ones to a vector and then delete them separately.

Note that using `std::unordered_set` instead of `std::vector` for `referenced_nodes` will also make these queries O(1) instead of O(n), which is probably better (at the cost of a slightly more expensive insert).